### DISCUSSION TITLE
How retain value of an internal variable in memory between calls of the function?

### DISCUSSION THREAD

**Discussion by luizmac** [OpenModelica/OpenModelica#10824] [Q&A]
Hello,

I am creating a block which emulates a controller. Inside this block I am using a function call to implement my controller. This is a first step to further use this block for software-in-the-loop simulations with C code.

In my controller I need to retain a value of a variable to use in the next step (integrator) but I do not know how to implement this on modelica. Every call of the function all internal variable are reset. On simulink we have two options:  using a persistent variable or putting the desired variable as an output and feedback it with a memory block.

How can I retain these values to the next call in modelica?

My code is : 

model State_Feedback_PI
  extends Modelica.Blocks.Icons.Block;
  Modelica.Blocks.Interfaces.RealInput x1;
  Modelica.Blocks.Interfaces.RealInput x2;
  Modelica.Blocks.Interfaces.RealInput referencia;
  Modelica.Blocks.Interfaces.RealOutput u;
  Modelica.Blocks.Interfaces.RealOutput test;

///////////////////////////////////////////////////////
parameter Real kx1;
parameter Real kx2;
parameter Real ki;
parameter Real c1;
parameter Real c2;
////////////////////////////////////////

/////////////////////////////////////////////////
/////////////////////////////////////////////////
function fn

  input  Real     x1;
  input  Real     x2;
  input  Real     kx1;
  input  Real     kx2;
  input  Real     ki;
  input  Real     c1;
  input  Real     c2;
  input  Real     referencia;
  output Real     u;
  output Real     test;
  
/////////////////////////////////////////////////

protected  
Real erro;
Real csi;
Real u_ant;

///////////////////////////////////////////////////////////////////

algorithm  

test := u_ant;
erro := referencia-1*(c1*x1+c2*x2);
csi  := 1;
u    := -kx1*x1 - kx2*x2 - ki*csi;
u_ant:= u;

// Here notice that the value u_ant should be retained to the next call. The variable test should be a delayed version of u.

end fn;

  /////////////////////////////////////////////////////////////////
  
algorithm 
 (u,test) := fn(x1,x2,kx1,kx2,ki,c1,c2,referencia);

**Comment by casella** [comment]
Functions in Modelica are memoryless. Persistent variables are either continuous states, that depend on differential equations but can also be modified discontinuously via reinit) or discrete states, that are updated by when equations.

The memory block of Simulink is evil 😅 . There is no such thing in Modelica, it's not really a declarative concept.

Do you want to represent a discrete PI controller or a continous-time PI controller?

**Reply by luizmac** [reply]
Hi,

I want to create a function which executes my controller. It  has an discrete integrator. I trying to create a function which emulates the code presented in my microcontroller. There all the calculations are occurring inside a c file. And on this example I trying to use a function block instead a c one.

**Comment by luizmac** [comment]
Hi I have opted to use the external C code declaring static variables. With this solution I could write an simple algorithm that retain the values of the variable and use it  in the next function call. 

Now my doubt is: Is this the only solution? How can I write an algorithm using modelica language in order to retain the value for the next function call?   Is this a language limitation?

**Comment by casella** [comment]
> Hi I have opted to use the external C code declaring static variables. With this solution I could write an simple algorithm that retain the values of the variable and use it in the next function call.

This is really not recommended, unless you declare the function as impure. Modelica functions are expected to be memoryless, which means the compiler is free to perform optimizations based on the assumption that calling the same function twice with the same arguments will lead to the same result. If this is not true, the generated code could be plain wrong.
 
> Now my doubt is: Is this the only solution? How can I write an algorithm using modelica language in order to retain the value for the next function call? Is this a language limitation?

You need to use discrete variables. For example, this is how you can implement a first-order SISO discrete-time system
```modelica
model DT
  discrete Real x(start = x_start, fixed = true);
  input Real u;
  output Real y;
  parameter Modelica.Units.SI.Time Ts = 1;
  parameter Real x_start = 0;
equation
  when sample(0, Ts) then
    x = f(pre(x), u);
    y = g(pre(x), u);
  end when
end DT;
```
f() and g() can be external functions written in C. But they are memoryless. The "state memory" is contained in the discrete variable x, and the compiler is fully aware of it. Otherwise it's a very dangerous hack.

**Reply by luizmac** [reply]
Thanks for the explanation and the example! I will declare my function as impure in order to avoid any problems. 

My idea comes from: 

[https://mbe.modelica.university/behavior/functions/controller/#sil-controller]

But as you mentioned, the author also uses impure functions in order to avoid any execution problem.

**Comment by casella** [comment]
OK. Just make sure you call that function from inside a when sample statement, otherwise there is no guarantee that it will only be called once for every time step.
